File: examples/python/qnn_german_credit.py
"""使用 ArcQML Python API 训练德国信用数据集 QNN。"""

from __future__ import annotations

import csv
from pathlib import Path

import numpy as np
import arcqml


QUBITS = 10
STATE_DIMENSION = 1 << QUBITS
PARAMETERS_PER_LAYER = 37
LAYERS = 1
EPOCHS = 5
BATCH_SIZE = 100
LEARNING_RATE = 0.01
SEED = 4
DATA_PATH = Path(__file__).resolve().parents[1] / "data" / "german_credit.csv"
FEATURE_COLUMNS = (
    "Account Balance",
    "Payment Status of Previous Credit",
    "Purpose",
    "Value Savings/Stocks",
    "Length of current employment",
    "Guarantors",
    "Most valuable available asset",
    "Concurrent Credits",
    "Type of apartment",
    "No of Credits at this Bank",
)


def load_data(path: Path) -> tuple[np.ndarray, np.ndarray]:
    """读取 CSV，生成角度特征和二分类标签。"""
    with path.open(encoding="utf-8", newline="") as file:
        rows = list(csv.DictReader(file))
    features = np.asarray(
        [[float(row[column]) for column in FEATURE_COLUMNS] for row in rows],
        dtype=np.float64,
    )
    labels = np.asarray(
        [1.0 - float(row["Creditability"]) for row in rows],
        dtype=np.float64,
    )
    minimum = features.min(axis=0)
    maximum = features.max(axis=0)
    features = np.pi / 2.0 + (features - minimum) / (maximum - minimum) * np.pi
    return features, labels


def encode_product_states(features: np.ndarray) -> np.ndarray:
    """将 RX(x) 后接 RZ(x) 的特征编码直接转换为乘积态。"""
    states = np.ones((len(features), 1), dtype=np.complex128)
    for qubit in range(QUBITS - 1, -1, -1):
        angles = features[:, qubit]
        local_states = np.column_stack(
            (
                np.exp(-0.5j * angles) * np.cos(0.5 * angles),
                -1j * np.exp(0.5j * angles) * np.sin(0.5 * angles),
            )
        )
        states = np.einsum("bi,bj->bij", states, local_states).reshape(len(features), -1)
    assert states.shape == (len(features), STATE_DIMENSION)
    return np.ascontiguousarray(states, dtype=np.complex128)


def append_ansatz_layer(circuit: arcqml.Circuit, values: np.ndarray) -> None:
    """添加与 benchmark 相同的 37 参数、52 门 ansatz 层。"""
    if values.shape != (PARAMETERS_PER_LAYER,):
        raise ValueError("each ansatz layer requires 37 parameters")
    for qubit in range(QUBITS):
        circuit.rx(angle=float(values[qubit]), qubit=qubit)
    for control, target in ((0, 1), (2, 3), (4, 5), (9, 8), (7, 6)):
        circuit.cnot(control=control, target=target)
    for qubit in range(1, 9):
        circuit.rx(angle=float(values[9 + qubit]), qubit=qubit)
    for control, target in ((1, 2), (3, 4), (8, 7), (6, 5)):
        circuit.cnot(control=control, target=target)
    for qubit in range(2, 8):
        circuit.rx(angle=float(values[16 + qubit]), qubit=qubit)
    for control, target in ((2, 3), (4, 5), (7, 6)):
        circuit.cnot(control=control, target=target)
    for parameter, qubit in ((24, 3), (25, 4), (26, 5), (27, 6)):
        circuit.ry(angle=float(values[parameter]), qubit=qubit)
    circuit.cnot(control=3, target=4)
    circuit.cnot(control=6, target=5)
    circuit.rz(angle=float(values[28]), qubit=4)
    circuit.ry(angle=float(values[29]), qubit=4)
    circuit.rz(angle=float(values[30]), qubit=4)
    circuit.rz(angle=float(values[31]), qubit=5)
    circuit.ry(angle=float(values[32]), qubit=5)
    circuit.rz(angle=float(values[33]), qubit=5)
    circuit.cnot(control=4, target=5)
    circuit.rz(angle=float(values[34]), qubit=5)
    circuit.ry(angle=float(values[35]), qubit=5)
    circuit.rz(angle=float(values[36]), qubit=5)


def build_circuit(rng: np.random.Generator) -> arcqml.Circuit:
    circuit = arcqml.Circuit(num_qubits=QUBITS)
    for _layer in range(LAYERS):
        append_ansatz_layer(
            circuit=circuit,
            values=rng.standard_normal(PARAMETERS_PER_LAYER),
        )
    return circuit


def predict(circuit: arcqml.Circuit, observable: arcqml.PauliSum, states: np.ndarray) -> np.ndarray:
    simulator = arcqml.BatchStateVectorSimulator.from_amplitudes(
        num_qubits=QUBITS,
        amplitudes=states,
    )
    with arcqml.no_grad():
        return simulator.run(circuit=circuit, observable=observable).numpy()


def validate(
    circuit: arcqml.Circuit,
    observable: arcqml.PauliSum,
    states: np.ndarray,
    labels: np.ndarray,
) -> tuple[float, float]:
    scores = predict(circuit=circuit, observable=observable, states=states)
    loss = arcqml.binary_cross_entropy_with_logits(
        logits=arcqml.tensor(scores),
        targets=arcqml.tensor(labels),
    ).item()
    accuracy = np.mean((scores >= 0.0) == (labels == 1.0))
    return float(loss), float(accuracy)


def auc_metrics(scores: np.ndarray, labels: np.ndarray) -> tuple[float, float]:
    """不依赖 scikit-learn 计算 ROC-AUC 与梯形 PR-AUC。"""
    order = np.argsort(-scores, kind="stable")
    positives = float(labels.sum())
    negatives = float(len(labels) - labels.sum())
    tp = fp = roc_auc = pr_auc = 0.0
    previous_tpr = previous_fpr = previous_recall = 0.0
    previous_precision = 1.0
    for index in order:
        if labels[index] == 1.0:
            tp += 1.0
        else:
            fp += 1.0
        tpr, fpr = tp / positives, fp / negatives
        precision = tp / (tp + fp)
        roc_auc += (fpr - previous_fpr) * (tpr + previous_tpr) / 2.0
        pr_auc += (tpr - previous_recall) * (precision + previous_precision) / 2.0
        previous_tpr, previous_fpr = tpr, fpr
        previous_recall, previous_precision = tpr, precision
    return roc_auc, pr_auc


def main() -> None:
    features, labels = load_data(path=DATA_PATH)
    rng = np.random.default_rng(seed=SEED)
    indices = rng.permutation(len(features))
    train, validation, test = indices[:800], indices[800:900], indices[900:]
    states = encode_product_states(features=features)
    circuit = build_circuit(rng=rng)
    observable = arcqml.PauliSum.z(num_qubits=QUBITS, qubit=5)
    optimizer = arcqml.Adam(learning_rate=LEARNING_RATE)

    print(f"samples: train={len(train)}, validation={len(validation)}, test={len(test)}")
    print(f"qubits={QUBITS}, layers={LAYERS}, parameters={circuit.num_parameters}")
    for epoch in range(1, EPOCHS + 1):
        rng.shuffle(train)
        weighted_loss = 0.0
        for start in range(0, len(train), BATCH_SIZE):
            batch_indices = train[start : start + BATCH_SIZE]
            simulator = arcqml.BatchStateVectorSimulator.from_amplitudes(
                num_qubits=QUBITS,
                amplitudes=np.ascontiguousarray(states[batch_indices]),
            )
            logits = simulator.run(circuit=circuit, observable=observable)
            targets = arcqml.tensor(np.ascontiguousarray(labels[batch_indices]))
            loss = arcqml.binary_cross_entropy_with_logits(logits=logits, targets=targets)
            loss.backward()
            optimizer.step(circuit=circuit)
            optimizer.zero_grad(circuit=circuit)
            weighted_loss += float(loss.item()) * len(batch_indices)

        validation_loss, validation_accuracy = validate(
            circuit=circuit,
            observable=observable,
            states=np.ascontiguousarray(states[validation]),
            labels=np.ascontiguousarray(labels[validation]),
        )
        print(
            f"epoch {epoch:02}/{EPOCHS}: train_loss={weighted_loss / len(train):.6f}, "
            f"validation_loss={validation_loss:.6f}, "
            f"validation_accuracy={validation_accuracy * 100:.1f}%"
        )

    test_scores = predict(
        circuit=circuit,
        observable=observable,
        states=np.ascontiguousarray(states[test]),
    )
    roc_auc, pr_auc = auc_metrics(scores=test_scores, labels=labels[test])
    print(f"test ROC-AUC={roc_auc:.4f}, PR-AUC={pr_auc:.4f}")


if __name__ == "__main__":
    main()

File: examples/rust/qnn_german_credit.rs
use arcqml::prelude::*;
use num_complex::Complex64;
use std::{cmp::Ordering, f64::consts::PI, fs};
use arcqml_visualization::write_svg;

type AppResult<T> = std::result::Result<T, Box<dyn std::error::Error>>;

const QUBITS: usize = 10;
const STATE_DIMENSION: usize = 1 << QUBITS;
const PARAMETERS_PER_LAYER: usize = 37;
const LAYERS: usize = 1;
const EPOCHS: usize = 5;
const BATCH_SIZE: usize = 100;
const LEARNING_RATE: f64 = 0.01;
const SEED: u64 = 4;
const DATA_PATH: &str = "examples/data/german_credit.csv";
const FEATURE_COLUMNS: [&str; QUBITS] = [
    "Account Balance",
    "Payment Status of Previous Credit",
    "Purpose",
    "Value Savings/Stocks",
    "Length of current employment",
    "Guarantors",
    "Most valuable available asset",
    "Concurrent Credits",
    "Type of apartment",
    "No of Credits at this Bank",
];

#[derive(Clone)]
struct Sample {
    features: [f64; QUBITS],
    label: f64,
}

struct Split {
    train: Vec<Sample>,
    validation: Vec<Sample>,
    test: Vec<Sample>,
}

struct Rng(u64);

impl Rng {
    fn new(seed: u64) -> Self {
        Self(seed)
    }

    fn next_u64(&mut self) -> u64 {
        self.0 = self
            .0
            .wrapping_mul(6_364_136_223_846_793_005)
            .wrapping_add(1_442_695_040_888_963_407);
        self.0
    }

    fn uniform(&mut self) -> f64 {
        ((self.next_u64() >> 11) as f64 + 0.5) / ((1u64 << 53) as f64)
    }

    fn normal(&mut self) -> f64 {
        let radius = (-2.0 * self.uniform().ln()).sqrt();
        radius * (2.0 * PI * self.uniform()).cos()
    }

    fn shuffle<T>(&mut self, values: &mut [T]) {
        for index in (1..values.len()).rev() {
            let selected = (self.next_u64() as usize) % (index + 1);
            values.swap(index, selected);
        }
    }
}

fn load_data(path: &str) -> AppResult<Split> {
    let text = fs::read_to_string(path)?;
    let mut lines = text.lines();
    let headers: Vec<&str> = lines.next().ok_or("CSV is empty")?.split(',').collect();
    let label_column = headers
        .iter()
        .position(|name| *name == "Creditability")
        .ok_or("missing Creditability column")?;
    let feature_indices: Vec<usize> = FEATURE_COLUMNS
        .iter()
        .map(|wanted| {
            headers
                .iter()
                .position(|name| name == wanted)
                .ok_or_else(|| format!("missing feature column: {wanted}"))
        })
        .collect::<std::result::Result<Vec<_>, String>>()?;

    let mut samples = Vec::new();
    for (row_index, line) in lines.enumerate() {
        let fields: Vec<&str> = line.split(',').collect();
        if fields.len() != headers.len() {
            return Err(format!("invalid CSV row {}", row_index + 2).into());
        }
        let mut features = [0.0; QUBITS];
        for (output, &column) in features.iter_mut().zip(&feature_indices) {
            *output = fields[column].parse()?;
        }
        // 原始数据：1=信用良好、0=信用不良；基准将“信用不良”作为正类。
        let creditability: u8 = fields[label_column].parse()?;
        samples.push(Sample {
            features,
            label: f64::from(1 - creditability),
        });
    }

    // 与 benchmark 预处理相同，将每列 min-max 缩放到 [π/2, 3π/2]。
    for column in 0..QUBITS {
        let minimum = samples
            .iter()
            .map(|sample| sample.features[column])
            .fold(f64::INFINITY, f64::min);
        let maximum = samples
            .iter()
            .map(|sample| sample.features[column])
            .fold(f64::NEG_INFINITY, f64::max);
        for sample in &mut samples {
            sample.features[column] =
                PI / 2.0 + (sample.features[column] - minimum) / (maximum - minimum) * PI;
        }
    }

    Rng::new(SEED).shuffle(&mut samples);
    let test = samples.split_off(900);
    let validation = samples.split_off(800);
    Ok(Split {
        train: samples,
        validation,
        test,
    })
}

fn encode_product_states(samples: &[Sample]) -> AppResult<Tensor> {
    let mut amplitudes = Vec::with_capacity(samples.len() * STATE_DIMENSION);
    for sample in samples {
        let mut state = vec![Complex64::new(1.0, 0.0)];
        for qubit in (0..QUBITS).rev() {
            let angle = sample.features[qubit];
            let half = angle / 2.0;
            let phase_zero = Complex64::from_polar(1.0, -half) * half.cos();
            let phase_one =
                Complex64::new(0.0, -1.0) * Complex64::from_polar(1.0, half) * half.sin();
            let mut expanded = Vec::with_capacity(state.len() * 2);
            for amplitude in state {
                expanded.push(amplitude * phase_zero);
                expanded.push(amplitude * phase_one);
            }
            state = expanded;
        }
        amplitudes.extend(state);
    }
    Ok(Tensor::new(TensorData::FlatC64 {
        data: amplitudes,
        shape: vec![samples.len(), STATE_DIMENSION],
    })?)
}

fn append_ansatz_layer(circuit: &mut Circuit, values: &[f64]) -> AppResult<()> {
    if values.len() != PARAMETERS_PER_LAYER {
        return Err("each ansatz layer requires 37 parameters".into());
    }
    for qubit in 0..QUBITS {
        circuit.rx(values[qubit], qubit)?;
    }
    for (control, target) in [(0, 1), (2, 3), (4, 5), (9, 8), (7, 6)] {
        circuit.cnot(control, target)?;
    }
    for qubit in 1..9 {
        circuit.rx(values[9 + qubit], qubit)?;
    }
    for (control, target) in [(1, 2), (3, 4), (8, 7), (6, 5)] {
        circuit.cnot(control, target)?;
    }
    for qubit in 2..8 {
        circuit.rx(values[16 + qubit], qubit)?;
    }
    for (control, target) in [(2, 3), (4, 5), (7, 6)] {
        circuit.cnot(control, target)?;
    }
    for (parameter, qubit) in [(24, 3), (25, 4), (26, 5), (27, 6)] {
        circuit.ry(values[parameter], qubit)?;
    }
    circuit.cnot(3, 4)?.cnot(6, 5)?;
    circuit
        .rz(values[28], 4)?
        .ry(values[29], 4)?
        .rz(values[30], 4)?;
    circuit
        .rz(values[31], 5)?
        .ry(values[32], 5)?
        .rz(values[33], 5)?;
    circuit.cnot(4, 5)?;
    circuit
        .rz(values[34], 5)?
        .ry(values[35], 5)?
        .rz(values[36], 5)?;
    Ok(())
}

fn build_circuit() -> AppResult<Circuit> {
    let mut rng = Rng::new(SEED);
    let mut circuit = Circuit::new(QUBITS)?;
    for _layer in 0..LAYERS {
        let values: Vec<f64> = (0..PARAMETERS_PER_LAYER).map(|_| rng.normal()).collect();
        append_ansatz_layer(&mut circuit, &values)?;
    }
    Ok(circuit)
}

fn labels(samples: &[Sample]) -> AppResult<Tensor> {
    Ok(Tensor::new(
        samples
            .iter()
            .map(|sample| sample.label)
            .collect::<Vec<_>>(),
    )?)
}

fn loss_and_backward(
    circuit: &Circuit,
    observable: &SparsePauliOp,
    samples: &[Sample],
) -> AppResult<f64> {
    let states = encode_product_states(samples)?;
    let simulator = BatchStateVectorSimulator::from_state_tensor(QUBITS, states)?;
    let logits = simulator.run(circuit, observable)?;
    let loss = binary_cross_entropy_with_logits_loss(&logits, &labels(samples)?)?;
    let value = loss.value()?;
    loss.backward()?;
    Ok(value)
}

fn predict(
    circuit: &Circuit,
    observable: &SparsePauliOp,
    samples: &[Sample],
) -> AppResult<Vec<f64>> {
    let _guard = no_grad();
    let states = encode_product_states(samples)?;
    let simulator = BatchStateVectorSimulator::from_state_tensor(QUBITS, states)?;
    let logits = simulator.run(circuit, observable)?;
    let storage = logits.storage();
    match &*storage {
        Storage::F64(values) => Ok(values.clone()),
        _ => Err("QNN logits must be F64".into()),
    }
}

fn validation(
    circuit: &Circuit,
    observable: &SparsePauliOp,
    samples: &[Sample],
) -> AppResult<(f64, f64)> {
    let scores = predict(circuit, observable, samples)?;
    let targets = labels(samples)?;
    let logits = Tensor::new(scores.clone())?;
    let loss = binary_cross_entropy_with_logits_loss(&logits, &targets)?.value()?;
    let correct = scores
        .iter()
        .zip(samples)
        .filter(|(score, sample)| (**score >= 0.0) == (sample.label == 1.0))
        .count();
    Ok((loss, correct as f64 / samples.len() as f64))
}

fn auc_metrics(scores: &[f64], samples: &[Sample]) -> (f64, f64) {
    let mut pairs: Vec<(f64, bool)> = scores
        .iter()
        .zip(samples)
        .map(|(&score, sample)| (score, sample.label == 1.0))
        .collect();
    pairs.sort_by(|left, right| right.0.partial_cmp(&left.0).unwrap_or(Ordering::Equal));
    let positives = pairs.iter().filter(|pair| pair.1).count() as f64;
    let negatives = pairs.len() as f64 - positives;
    let (mut tp, mut fp) = (0.0, 0.0);
    let (mut previous_tpr, mut previous_fpr, mut previous_recall, mut previous_precision) =
        (0.0, 0.0, 0.0, 1.0);
    let (mut roc_auc, mut pr_auc) = (0.0, 0.0);
    for (_, positive) in pairs {
        if positive {
            tp += 1.0;
        } else {
            fp += 1.0;
        }
        let tpr = tp / positives;
        let fpr = fp / negatives;
        let precision = tp / (tp + fp);
        roc_auc += (fpr - previous_fpr) * (tpr + previous_tpr) / 2.0;
        pr_auc += (tpr - previous_recall) * (precision + previous_precision) / 2.0;
        (previous_tpr, previous_fpr) = (tpr, fpr);
        (previous_recall, previous_precision) = (tpr, precision);
    }
    (roc_auc, pr_auc)
}

fn main() -> AppResult<()> {
    let data = load_data(DATA_PATH)?;
    let circuit = build_circuit()?;
    let observable = SparsePauliOp::single(QUBITS, 5usize, Pauli::Z, 1.0)?;
    let mut optimizer = Adam::new(LEARNING_RATE, 0.9, 0.999, 1e-8, 0.0)?;
    let mut rng = Rng::new(SEED);

    println!(
        "samples: train={}, validation={}, test={}",
        data.train.len(),
        data.validation.len(),
        data.test.len()
    );
    println!(
        "qubits={QUBITS}, layers={LAYERS}, parameters={}",
        circuit.num_parameters()
    );
    for epoch in 1..=EPOCHS {
        let mut indices: Vec<usize> = (0..data.train.len()).collect();
        rng.shuffle(&mut indices);
        let mut weighted_loss = 0.0;
        for batch_indices in indices.chunks(BATCH_SIZE) {
            let batch: Vec<Sample> = batch_indices
                .iter()
                .map(|&index| data.train[index].clone())
                .collect();
            let loss = loss_and_backward(&circuit, &observable, &batch)?;
            optimizer.step(circuit.parameters())?;
            optimizer.zero_grad(circuit.parameters());
            weighted_loss += loss * batch.len() as f64;
        }
        let (validation_loss, validation_accuracy) =
            validation(&circuit, &observable, &data.validation)?;
        println!(
            "epoch {epoch:02}/{EPOCHS}: train_loss={:.6}, validation_loss={validation_loss:.6}, validation_accuracy={:.1}%",
            weighted_loss / data.train.len() as f64,
            validation_accuracy * 100.0
        );
    }

    write_svg(&circuit, "./examples/rust/qnn_circuit.svg");

    let test_scores = predict(&circuit, &observable, &data.test)?;
    let (roc_auc, pr_auc) = auc_metrics(&test_scores, &data.test);
    println!("test ROC-AUC={roc_auc:.4}, PR-AUC={pr_auc:.4}");
    Ok(())
}
